lti_consumer/lti.py

```diff
@@ -0,0 +1,281 @@
+"""
+This module encapsulates code which implements the LTI specification.
+
+For more details see:
+https://www.imsglobal.org/activity/learning-tools-interoperability
+"""
+
+import logging
+import urllib
+import json
+
+from .exceptions import LtiError
+from .oauth import get_oauth_request_signature, verify_oauth_body_signature
+
+
+log = logging.getLogger(__name__)
+
+
+def parse_result_json(json_str):
+    """
+    Helper method for verifying LTI 2.0 JSON object contained in the body of the request.
+
+    The json_str must be loadable.  It can either be an dict (object) or an array whose first element is an dict,
+    in which case that first dict is considered.
+    The dict must have the "@type" key with value equal to "Result",
+    "resultScore" key with value equal to a number [0, 1],
+    The "@context" key must be present, but we don't do anything with it.  And the "comment" key may be
+    present, in which case it must be a string.
+
+    Arguments:
+        json_str (unicode):  The body of the LTI 2.0 results service request, which is a JSON string
+
+    Returns:
+        (float, str):  (score, [optional]comment) if verification checks out
+
+    Raises:
+        LtiError: if verification fails
+    """
+    try:
+        json_obj = json.loads(json_str)
+    except (ValueError, TypeError):
+        msg = "Supplied JSON string in request body could not be decoded: {}".format(json_str)
+        log.info("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # the standard supports a list of objects, who knows why. It must contain at least 1 element, and the
+    # first element must be a dict
+    if not isinstance(json_obj, dict):
+        if isinstance(json_obj, list) and len(json_obj) >= 1 and isinstance(json_obj[0], dict):
+            json_obj = json_obj[0]
+        else:
+            msg = ("Supplied JSON string is a list that does not contain an object as the first element. {}"
+                   .format(json_str))
+            log.info("[LTI] %s", msg)
+            raise LtiError(msg)
+
+    # '@type' must be "Result"
+    result_type = json_obj.get("@type")
+    if result_type != "Result":
+        msg = "JSON object does not contain correct @type attribute (should be 'Result', is z{})".format(result_type)
+        log.info("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # '@context' must be present as a key
+    REQUIRED_KEYS = ["@context"]  # pylint: disable=invalid-name
+    for key in REQUIRED_KEYS:
+        if key not in json_obj:
+            msg = "JSON object does not contain required key {}".format(key)
+            log.info("[LTI] %s", msg)
+            raise LtiError(msg)
+
+    # 'resultScore' is not present.  If this was a PUT this means it's actually a DELETE according
+    # to the LTI spec.  We will indicate this by returning None as score, "" as comment.
+    # The actual delete will be handled by the caller
+    if "resultScore" not in json_obj:
+        return None, json_obj.get('comment', "")
```

Comment:
I'm confused, if result score is not present and this is a PUT request it's actually a delete, but this code does not check for the the HTTP Verb. Question, if this method is parsing the JSON, should this piece of the logic be in this method at all?